http-webhook.md

```diff
@@ -0,0 +1,339 @@
+processing status is unknown, the response MUST have the [202 Accepted][202]
+status code.
+
+If the delivery cannot be accepted because the notification format has not
+been understood, the service MUST respond with status code [415][415].
+
+All further error status codes apply as specified in [RFC7231][RFC7231].
+
+## 3. Authorization
+
+The delivery request MUST use one of the following two methods, both of which
+lean on the OAuth 2.0 Bearer Token [RFC6750][RFC6750] model.
+
+The delivery target MUST support both methods.
+
+The client MAY use any token-based authorization scheme. Challenge-based
+schemes MUST NOT be used.
+
+### 3.1.  Authorization Request Header Field
+
+The access token is sent in the [`Authorization`][Authorization] request
+header field defined by HTTP/1.1.
+
+For [OAuth 2.0 Bearer][Bearer] tokens, the "Bearer" scheme MUST be used.
+
+Example:
+
+``` text
+POST /resource HTTP/1.1
+Host: server.example.com
+Authorization: Bearer mF_9.B5f-4.1JqM
+```
+
+### 3.2 URI Query Parameter
+
+When sending the access token in the HTTP request URI, the client adds the
+access token to the request URI query component as defined by "Uniform Resource
+Identifier (URI): Generic Syntax" [RFC3986][RFC3986], using the "access_token"
+parameter.
+
+For example, the client makes the following HTTP request:
+
+``` text
+POST /resource?access_token=mF_9.B5f-4.1JqM HTTP/1.1
+Host: server.example.com
+```
+
+The HTTP request URI query can include other request-specific parameters, in
+which case the "access_token" parameter MUST be properly separated from the
+request-specific parameters using "&" character(s) (ASCII code 38).
+
+For example:
+
+    https://server.example.com/resource?access_token=mF_9.B5f-4.1JqM&p=q
+
+Clients using the URI Query Parameter method SHOULD also send a
+Cache-Control header containing the "no-store" option.  Server
+success (2XX status) responses to these requests SHOULD contain a
+Cache-Control header with the "private" option.
+
+Because of the security weaknesses associated with the URI method (see
+[RFC6750, Section 5][RFC6750]), including the high likelihood that the URL
+containing the access token will be logged, it SHOULD NOT be used unless it is
+impossible to transport the access token in the "Authorization" request header
+field or the HTTP request entity-body. All further caveats cited in
+[RFC6750][RFC6750] apply equivalently.
+
+## 4. Abuse Protection
+
+Any system that allows registration of and delivery of notifications to
+arbitary HTTP endpoints can potentially be abused such that someone maliciously
+or inadvertently registers the address of a system that does not expect such
+requests and for which the registering party is not authorized to perform such
+a registration. In extreme cases, a notification infrastructure could be abused
+to launch denial-of-service attacks against an arbitrary web-site.
+
+To protect the sender from being abused in such a way, a legitimate delivery
```

Comment:
Across the industry (facebook, dropbox, slack, etc.) the most common (and much simpler) approach for URL validation is to send a challenge. I realize you don't want to use a challenge for authorization, but I believe it is perfectly acceptable for validation. If the URL is invalid, or the delivery target doesn't want to accept a webhook call, it simply doesn't respond to the initial challenge, and in that case the sender MUST NOT deliver events to the target.

Reply:
The point here is that we need a standard way to express the opt-in. Everyone does these thibgs, but each in their own way. The opt-in can be a reply or the opt-in can be a callback to a URL; either one is fine and we can also offer both.

Reply:
@rperelma I just added a callback option.

Reply:
@clemensv thanks, but not quite what I had in mind :)
I was referring to a simple challenge sent to the target, and the target responding (or not) to that challenge, thereby accepting (or not) further deliveries. This is a very widely supported approach across the industry.

Reply:
Facebook has this for example: https://developers.facebook.com/docs/graph-api/webhooks/getting-started see the section on verification.

Reply:
@rperelma that's exactly what this mechanism is, though. The client sends a string (WebHook-Request-Origin) and the server echoes that string (WebHook-Allowed-Origin) if and only if it's ok with that source pushing. I do that on OPTIONS, because that's what OPTIONS is for (also see W3C CORS). The rate negotiation is optional.

What I believe I need for a multi-party interop standard is a way for the target to not only echo back some arbitrary string, but for the target to also have a way to make semantic sense out of that string, because the webhook might not just be written for a single client system, but for multiple different client systems and the target may want to be able to differentiate between those.

The origin challenge might be "instance-a35f7d9e.service.example.com" and the target may have a rule to permit "*.example.com" and then grants access by echoing the origin phrase. The implementation of the target might also be ok with anyone pushing to them and then just replies with "*" out of a dumb response rule. 

If I could mandate a tighter authorization model (which I would if this were a product and not a standards design), I would even tie the Origin into the access token. I can't do that here, because there's appears to be majority who work with simple symmetric keys.